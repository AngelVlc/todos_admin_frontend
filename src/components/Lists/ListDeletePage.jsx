import React from 'react';
import { useHistory } from "react-router-dom";

export const ListDeletePage = () => {
    let history = useHistory();

    return (
        <div className="container">
            <h3 className="title">LIST DELETE PAGE</h3>
            <button className="button" onClick={() => history.push('/')}>Home</button>
        </div>
    )
}
