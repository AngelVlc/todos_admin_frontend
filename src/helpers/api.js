import { requestPending, requestDone, requestFailed } from '../actions'

const backendUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:5001'

export const doGetToken = (loginDto, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(loginDto)
        };
        fetch(`${backendUrl}/auth/token`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (!res.ok) {
                    res.text().then(txt => reject(txt));
                    return
                }
                res.json().then(obj => {
                    resolve(obj);
                })
            })
            .catch(error => {
                requestsDispatch(requestDone());
                reject(error.message);
            })
    });
}

const getHeaders = (method, token) => {
    return {
        method: method,
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` },
    };
}

export const doGet = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('GET', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    res.json().then(obj => {
                        resolve(obj);
                    })
                } else {
                    res.text().then(txt => {
                        requestsDispatch(requestFailed(txt));
                        reject(txt)
                    });
                }
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doDelete = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('DELETE', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    resolve();
                    return
                }
                res.text().then(txt => {
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doPost = (endpoint, body, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('POST', token);
        requestOptions.body = JSON.stringify(body);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                res.text().then(txt => {
                    if (res.ok) {
                        resolve(txt);
                        return
                    }
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}
