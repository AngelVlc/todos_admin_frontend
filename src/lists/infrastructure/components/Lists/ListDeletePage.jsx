import React, { useEffect, useState } from 'react';
import { Link, useParams, useHistory } from 'react-router-dom';
import axios from 'axios';

export const ListDeletePage = () => {
    const [list, setList] = useState(null);
    let { listId } = useParams();
    let history = useHistory();

    useEffect(() => {
        const getList = async () => {
            const res = await axios.get(`lists/${listId}`)
            const info = {
                ...res.data,
                title: `Delete list ${res.data.name}`
            }
            setList(info);
        }
            getList();
    }, [listId]);

    const deleteList = async () => {
        await axios.delete(`lists/${listId}`)
        history.push('/lists');
    }

    return (
        <>
            {list &&
                <div className="container">
                    <h3 className="title">{list.title}</h3>
                    <nav className="breadcrumb" aria-label="breadcrumbs">
                        <ul>
                            <li><Link to={`/`}>Home</Link></li>
                            <li><Link to={`/lists`}>Lists</Link></li>
                            <li><Link to={`/lists/${listId}/edit`}>List</Link></li>
                            <li className="is-active"><Link aria-current="page" to={`/lists/${listId}/delete`}>{`Delete ${list.name}`}</Link></li>
                        </ul>
                    </nav>
                    <div className="buttons">
                        <button className="button is-danger" data-testid="yes" onClick={() => deleteList()}>YES</button>
                        <button className="button" data-testid="no" onClick={() => history.goBack()}>NO</button>
                    </div>
                </div>
            }
        </>
    );
}
