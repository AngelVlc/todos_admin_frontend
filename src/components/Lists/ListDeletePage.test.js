import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { ListDeletePage } from './ListDeletePage'
import { AppContext } from '../../contexts/AppContext'
import * as api from '../../helpers/api';
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils';

jest.mock('../../helpers/api');
const mockHistoryGoBack = jest.fn();
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        goBack: mockHistoryGoBack,
        push: mockHistoryPush
    })
}));

const renderWithContextAndRouter = (component) => {
    api.doGet.mockResolvedValue({
        id: 2,
        name: 'ListName'
    });
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/lists/2/delete`]}>
                    <Route path="/lists/:listId/delete">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should match the snapshot', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouter(<ListDeletePage />);
        fragment = asFragment;
    });
    expect(fragment(<ListDeletePage />)).toMatchSnapshot();
});

it('should cancel the deletion', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouter(<ListDeletePage />);
    });

    await wait(() => {
        fireEvent.click(container.getByTestId('no'));
    })

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should delete the List', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouter(<ListDeletePage />);
    });

    api.doDelete.mockResolvedValue({});

    await wait(() => {
        fireEvent.click(container.getByTestId('yes'));
    })

    expect(api.doDelete.mock.calls.length).toBe(1);
    expect(api.doDelete.mock.calls[0][0]).toBe('lists/2');
    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/lists');
    mockHistoryPush.mockClear();
});
