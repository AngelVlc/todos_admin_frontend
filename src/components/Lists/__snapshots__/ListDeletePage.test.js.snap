// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Delete list ListName
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/lists"
          >
            Lists
          </a>
        </li>
        <li>
          <a
            href="/lists/2/edit"
          >
            List
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/lists/2/delete"
          >
            Delete ListName
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="yes"
      >
        YES
      </button>
      <button
        class="button"
        data-testid="no"
      >
        NO
      </button>
    </div>
  </div>
</DocumentFragment>
`;
